proposed/2022/dotnet-nuget-why-proposal.md

```diff
@@ -0,0 +1,140 @@
+
+## Goals
+The main goal is to add `dotnet nuget why` command to .NET CLI to help customers understand the dependency graph of a given package for PackageReference projects.
+
+Projects that use `packages.config` format are out of scope for this effort because the XML file contains a flat list of both direct and transitive package dependencies. Customers cannot visualize the dependency graph from the file itself but it would be easy to update the package version of transitive depedency incase if there is any security vulnerability in the current version.
+
+## Customers
+We are building this for .NET users to be successful with understanding dependency graph for a given package.
+
+## Solution
+NuGet restore operation generates following files under `obj` folder for `PackageReference` style projects.
+- `project.assets.json` file that maintains a project's dependency graph, which is used to make sure that all necessary packages are installed on the computer.
+- `{projectName}.nuget.dgspec.json` file that maintains a project's top-level dependencies along with other metadata.
+- `{projectName}nuget.g.targets` and `{projectName}.nuget.g.props` files define various properties, such as the path to where the packages cache is on the local machine. They also include any MSBuild imports that NuGet packages referenced by the project need.
+
+The contents of `{projectName}.nuget.dgspec.json` and `project.assets.json` will be helpful to understand the dependency graph of a given package but there are few limitations.
+
+-  NuGet restore operation downloads packages during dependency resolution which may not part of the final graph The absence of a package id and version combination in `project.assets.json` file signals that it was downloaded during dependency resolution but not part of the final dependency graph.
+- Packages acquired through `PackageDownload` are not tied to the project in any way beyond acquisition. These packages are not recorded in the final graph that `project.assets.json` maintains.
+
+`dotnet nuget why` command prints the dependency graph of a given package only if it is part of the final graph.
+
+```
+dotnet nuget why [<PROJECT>|<SOLUTION>] package <PACKAGE_NAME>
+    [-f|--framework <FRAMEWORK>]
+    [--version <VERSION>]
+    [-v|--verbosity <LEVEL>]
+
+dotnet nuget why -h|--help
+```
+#### Arguments
+
+- PROJECT | SOLUTION
+
+The project or solution file to operate on. If not specified, the command searches the current directory for one. If more than one solution or project is found, an error is thrown.
+
+- PACKAGE_NAME
+
+The package for whom the dependency graph has to be identified. Package name wildcards are not supported.
+
+#### Options
+
+- -f|--framework <FRAMEWORK>
+
+Search package depdnency graph only for a specific target framework.
+
+- -?|-h|--help
+
+Prints out a description of how to use the command.
+
+- --version <VERSION>
+
+Version of the package. See NuGet package versioning.
```

Comment:
Given that `why` will be used to show information about the resolved graph, and that NuGet only allows a single version of a package to be selected in the project, I don't understand the use-case for `--version`.  Lets say that PackageA version 1.2.3 is selected in the project, what happens if the customer does `--version 1.0.0`?

Reply:
Good question. `dotnet nuget why package` can be executed on a solution level just like `dotnet list package` command.

My understanding is that if multiple projects that are unrelated have different version of a same package installed then it would be convenient for customers to search for a specific version of a package. 

![image](https://user-images.githubusercontent.com/52756182/173682368-61801549-2f29-416f-8151-95dbbe752e05.png)

Reply:
> My understanding is that if multiple projects that are unrelated have different version of a same package installed then it would be convenient for customers to search for a specific version of a package.

I think we should start with a more minimal solution before we add things like version.

Reply:
Packages are managed on a project level. 

If running it on a solution, we just show project number of outputs. 

Running on a solution just means running it for each project separately.

Reply:
Makes sense, I agree we can remove the version part for the first minimal version.